modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java

```diff
@@ -988,6 +988,8 @@ public void rollbackUserIndex(String idxName) {
             GridH2IndexBase rmvIdx = tmpIdxs.remove(idxName);
 
             assert rmvIdx != null;
+
+            rmvIdx.destroy(true);
```

Comment:
Indexes creation is part of `SchemaManager#createIndex`:

```
final GridH2IndexBase h2Idx = desc.createUserIndex(idxDesc, cacheVisitor);
```

This method `rollbackUserIndex` is executed iff a SQL query statement fails (try-catch block). But test halts and failes earlier in `desc.createUserIndex`. 

Then, index isn't registered in table due to an exception in `FailBuildIndexConsumer`, and then it becomes an orphaned index. There is a method `IndexProcessor#destroyOrphanIndex` for such indexes. I see that it actually uses only by special command, but maybe it's possible to reuse this metod and clean such indexes on cache stop. 

WDYT?


, and also not cleaned along with tables in moment of a cache stop. It looks like a bug.

Reply:
Or, we can destroy index immediately on the closure failure. Because there is no chances to use this index later.

Reply:
I'm not sure dropping index only on closure failure will be helpful, because destroy operation may fail either.
  
I think we should destroy orphan indices on table start (on or after WAL recovery) as well, due to the index tree can't be reused by the rebuild operation. 

If we can't detect and drop orphan index tree on WAL recovery synchronously, then we should drop orphans on cache destroy as well, because cache could be dropped concurrently while recovery task isn't started yet.